### DISCUSSION TITLE
discussion concerning leave_add_csync.go

### DISCUSSION THREAD

**Discussion by romu42** [DNSSEC-Provisioning/music#151] [Q&A] [answered]
The leave_add_csync process names a `FSMSigner`, I am uncertain what this is and need a little help with the logic/implementation.

leave_add_csync.go

.... line 132-136
```
// Semantics:
// 1. Lookup zone signergroup (can only be one)
// 2. Lookup FSMSigner for the zone (can only be one)
// 3. Go through the steps below.
// 4. Celebrate Christmas
```
....

what is the `FSMSigner`?
how/when is it set?

**Comment by johanix** [comment]
Nu har jag kollat runt en hel del och min slutsats är att zone.FSMSigner är en rest från tidigare när zoner kunde ansluta sig (och koppla bort sig) från individuella signers. Se t ex funktionerna ZoneAttachFsm(..) och ZoneDetachFsm(): Den förra går att nå, men med en debug-kommentar. Den andra existerar men anropas ingenstans. Typisk leftover.

Numera så är signergroups själva korskopplingspunkten. Signers “joins/leaves” signergroups. Zoner “joins/leaves” signergroups. Och därmed så löste vi knuten att inte behöva hålla koll på alla signers i alla zoner.

**Comment by johanix** [comment, marked as answer]
Mer kollande tyder på att denna behövs just i leaveoperationerna. När en signer försöker "leave signergroup" så sätts samtliga zoner i denna signer group i REMOVE-SIGNER processen, signergruppen markerar denna signer som "pending removal" och alla zoner får denna signer som z.FSMsigner samt hela signergroup blir "locked". Locked innebär att inga zoner kan tas bort eller läggas till samt att inga signers kan tas bort eller läggas till. Dvs vi accepterar bara en "leave" transaktion åt gången.

I takt med att zonerna tickar igenom REMOVE-SIGNER så kommer de att ha behov av att veta vilken signer det är som är på väg bort. Det är vad de använder z.FSMsigner till. När en zon är klar med REMOVE-SIGNER så flaggar den detta till signergroup och när antalet zoner som är i REMOVE-SIGNER når noll så ändras signergroup.Locked till false.

Jag tror kort sagt att logiken är korrekt som den är, att z.FSMsigner behövs samt att ingen kod skall tas bort.
